### ISSUE TITLE
Optimizer Study 01

### ISSUE THREAD

**Issue by jaeminh** [jaeminh/notes#1]
## example code
```python
import torch
import torch.nn as nn

# Create a simple linear layer (weight matrix)
torch.manual_seed(0)  # For reproducibility
linear = nn.Linear(3, 2, bias=True)  # 3 input features, 2 output features

# Input data
x_input = torch.tensor([[1.0, 2.0, 3.0]], requires_grad=True)  # shape: (1, 3)

# Target (label)
target = torch.tensor([[0.5, 1.5]])  # shape: (1, 2)

# Forward pass
output = linear(x_input)
output.retain_grad()

# Compute L2 loss (Mean Squared Error)
loss = torch.norm(output - target, p=2)  # L2 norm of the difference
# loss.retain_grad()

# Backward pass
loss.backward()

print(f"Input: {x_input}")
print(f"Weight matrix: {linear.weight}, {linear.weight.shape}")
print(f"Bias: {linear.bias}, {linear.bias.shape}")
print(f"Output: {output}")
print(f"Target: {target}")
print(f"L2 Loss: {loss.item()}")
print()

print(f"Gradient ∂L/∂input: {x_input.grad}")
print(f"Manual ∂L/∂input: {torch.matmul(output.grad, linear.weight)}")
print()

print(f"Gradient ∂L/∂weight: {linear.weight.grad}")
print(f"Manual ∂L/∂weight: {torch.matmul(output.grad.t(), x_input)}")
print()

print(f"Gradient ∂L/∂bias: {linear.bias.grad}")
```

## torch.nn.Linear: $y = x{W}^{T} + b$
https://docs.pytorch.org/docs/stable/generated/torch.nn.Linear.html

## Jacobian
https://en.wikipedia.org/wiki/Jacobian_matrix_and_determinant
$f : \mathbb{R}^n \to \mathbb{R}^m$ 에서 $f$ 가 input vector $\mathbf{x} \in \mathbb{R}^m$ 에 대해 1차 편미분이 가능하다면 Jacobian matrix of f는 아래와 같다.
Jacobian matrix $J_f \in \mathbb{R}^{m \times n}$ 으로 $\text{output dim} \times \text{input dim}$

<img width="544" height="160" alt="Image" src="https://github.com/user-attachments/assets/416ee1f6-a52b-4b25-8ddf-b65e3d8da99f" />

## gradient with chain role
$\large \text{Linear}: y = xW^T + b$
$x \in \mathbb{R}^3$, $y \in \mathbb{R}^3$, $W \in \mathbb{R}^{2 \times 3}$, $b \in \mathbb{R}^2$ but skip bias

$$
\begin{cases}
y_1 = x_1 W_{11} + x_2 W_{12} + x_3 W_{13} \\
y_2 = x_1 W_{21} + x_2 W_{22} + x_3 W_{23}
\end{cases}
$$

### compute $\large \frac{\partial \mathbf{y}}{\partial \mathbf{x}}$
$\large \frac{\partial y_j}{\partial x_i} = W_{j,i}$ 이므로 Jacobian을 적용하면

$$
\large \frac{\partial \mathbf{y}}{\partial \mathbf{x}} 
= \begin{bmatrix} W_{11} & W_{12} & W_{13} \\ 
W_{21} & W_{22} & W_{23} \end{bmatrix} = W \in \mathbb{R}^{2 \times 3}
$$

### compute $\large \frac{\partial L}{\partial \mathbf{y}}$
Jacobian을 적용하면 $\large \frac{\partial L}{\partial \mathbf{y}} \in \mathbb{R}^{1 \times 2}$ 이므로 

$$
\large \begin{bmatrix} \frac{\partial L}{\partial y_1} & \frac{\partial L}{\partial y_2}\end{bmatrix}
$$

### compute $\large \frac{\partial L}{\partial \mathbf{x}}$
chain rule에 의해서 $\large \frac{\partial L}{\partial \mathbf{x}} = \frac{\partial L}{\partial \mathbf{y}} \cdot \frac{\partial \mathbf{y}}{\partial \mathbf{x}} = \frac{\partial L}{\partial \mathbf{y}} \cdot W$ 이고 $\large \frac{\partial L}{\partial \mathbf{y}} \in \mathbb{R}^{1 \times 2}$ , $\large W \in \mathbb{R}^{2 \times 3}$ 이므로 $\large \frac{\partial \mathbf{L}}{\partial \mathbf{x}} \in \mathbb{R}^{1 \times 3}$

## Conclusion
Linear layer의 gradient 계산을 위해 필요한 $\large \frac{\partial L}{\partial \mathbf{x}}$ 와  $\large \frac{\partial L}{\partial W}$ 는 아래와 같은 방법으로 구할 수 있다.
* $\large \frac{\partial L}{\partial \mathbf{x}} = \frac{\partial L}{\partial \mathbf{y}} \cdot W \in \mathbb{R}^{1 \times 3}$
```python
print(f"Gradient ∂L/∂input: {x_input.grad}")
print(f"Manual ∂L/∂input: {torch.matmul(output.grad, linear.weight)}")

# Gradient ∂L/∂input: tensor([[ 0.3167, -0.0447,  0.2061]])
# Manual ∂L/∂input: tensor([[ 0.3167, -0.0447,  0.2061]], grad_fn=<MmBackward0>)
```
* $\large \frac{\partial L}{\partial W} = (\frac{\partial L}{\partial \mathbf{y}})^{T} \cdot \mathbf{x} \in \mathbb{R}^{2 \times 3}$
```python
print(f"Gradient ∂L/∂weight: {linear.weight.grad}")
print(f"Manual ∂L/∂weight: {torch.matmul(output.grad.t(), x_input)}")

# Gradient ∂L/∂weight: tensor([[-0.6744, -1.3487, -2.0231],
#         [-0.7384, -1.4768, -2.2152]])
# Manual ∂L/∂weight: tensor([[-0.6744, -1.3487, -2.0231],
#         [-0.7384, -1.4768, -2.2152]], grad_fn=<MmBackward0>)
```
